File: app/RepoShelf.Core/Jobs.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace RepoShelf.Core;

/// <summary>
/// Background jobs: GitHub Stars import and refresh-all. Jobs run on threads,
/// persist progress in `import_jobs` for resume, and support cancellation
/// between units of work.
/// </summary>
public sealed class JobManager
{
    private const int FailureListCap = 100;
    private static string Now() => DateTime.UtcNow.ToString("O");

    private readonly Store _store;
    private readonly GitHubClient _gh;
    private readonly RepoService _repos;
    private readonly int _maxRateLimitWaitMs;
    private readonly ConcurrentDictionary<string, CancellationTokenSource> _running = new();

    public JobManager(Store store, GitHubClient gh, RepoService repos, int maxRateLimitWaitMs = 90_000)
    {
        _store = store;
        _gh = gh;
        _repos = repos;
        _maxRateLimitWaitMs = maxRateLimitWaitMs;
    }

    private JsonObject? GetRow(string id)
    {
        lock (_store.Sync)
        {
            using var cmd = _store.Conn.CreateCommand();
            cmd.CommandText = "SELECT * FROM import_jobs WHERE id = $id";
            cmd.Parameters.AddWithValue("$id", id);
            using var reader = cmd.ExecuteReader();
            if (!reader.Read())
            {
                return null;
            }
            string? S(string c) => reader.IsDBNull(reader.GetOrdinal(c)) ? null : reader.GetString(reader.GetOrdinal(c));
            long L(string c) => reader.GetInt64(reader.GetOrdinal(c));
            return new JsonObject
            {
                ["id"] = S("id"),
                ["type"] = S("type"),
                ["username"] = S("username"),
                ["status"] = S("status"),
                ["options"] = JsonNode.Parse(S("options") ?? "{}"),
                ["nextPage"] = L("next_page"),
                ["processed"] = L("processed"),
                ["added"] = L("added"),
                ["updated"] = L("updated"),
                ["failed"] = L("failed"),
                ["failures"] = JsonNode.Parse(S("failures") ?? "[]"),
                ["error"] = S("error"),
                ["createdAt"] = S("created_at"),
                ["finishedAt"] = S("finished_at"),
            };
        }
    }

    public JsonObject? GetJob(string id) => GetRow(id);

    public JsonArray ListJobs()
    {
        lock (_store.Sync)
        {
            var ids = new List<string>();
            using (var cmd = _store.Conn.CreateCommand())
            {
                cmd.CommandText = "SELECT id FROM import_jobs ORDER BY created_at DESC LIMIT 20";
                using var reader = cmd.ExecuteReader();
                while (reader.Read())
                {
                    ids.Add(reader.GetString(0));
                }
            }
            return new JsonArray(ids.Select(id => (JsonNode?)GetRow(id)).ToArray());
        }
    }

    private void Update(string id, Action<SqliteUpdate> fill)
    {
        lock (_store.Sync)
        {
            var u = new SqliteUpdate(_store.Conn, id);
            fill(u);
            u.Execute();
        }
    }

    private sealed class SqliteUpdate
    {
        private readonly Microsoft.Data.Sqlite.SqliteConnection _conn;
        private readonly string _id;
        private readonly List<string> _sets = new();
        private readonly List<object?> _values = new();

        public SqliteUpdate(Microsoft.Data.Sqlite.SqliteConnection conn, string id)
        {
            _conn = conn;
            _id = id;
        }

        public void Set(string column, object? value)
        {
            _sets.Add($"{column} = $p{_values.Count}");
            _values.Add(value);
        }

        public void Execute()
        {
            using var cmd = _conn.CreateCommand();
            cmd.CommandText = $"UPDATE import_jobs SET {string.Join(", ", _sets)} WHERE id = $id";
            for (var i = 0; i < _values.Count; i++)
            {
                cmd.Parameters.AddWithValue($"$p{i}", _values[i] ?? DBNull.Value);
            }
            cmd.Parameters.AddWithValue("$id", _id);
            cmd.ExecuteNonQuery();
        }
    }

    private string InsertJob(string type, string? username, JsonObject options)
    {
        var id = Guid.NewGuid().ToString();
        lock (_store.Sync)
        {
            using var cmd = _store.Conn.CreateCommand();
            cmd.CommandText = """
                INSERT INTO import_jobs (id, type, username, status, options, created_at)
                VALUES ($id, $type, $user, 'running', $options, $ts)
                """;
            cmd.Parameters.AddWithValue("$id", id);
            cmd.Parameters.AddWithValue("$type", type);
            cmd.Parameters.AddWithValue("$user", (object?)username ?? DBNull.Value);
            cmd.Parameters.AddWithValue("$options", options.ToJsonString());
            cmd.Parameters.AddWithValue("$ts", Now());
            cmd.ExecuteNonQuery();
        }
        return id;
    }

    public JsonObject StartStarsImport(string username, bool includeReadme = true)
    {
        var id = InsertJob("stars", username, new JsonObject { ["includeReadme"] = includeReadme });
        Launch(id, RunStarsImportAsync);
        return GetRow(id)!;
    }

    public JsonObject StartRefreshAll()
    {
        var id = InsertJob("refresh_all", null, new JsonObject());
        Launch(id, RunRefreshAllAsync);
        return GetRow(id)!;
    }

    /// <summary>Import GitHub repository URLs found in a Chromium bookmarks file.</summary>
    public (JsonObject? Job, string? Error) StartBookmarksImport(string? bookmarksPath = null)
    {
        var urls = ChromeBookmarks.ExtractRepoUrls(bookmarksPath, out var sourcePath, out var error);
        if (urls is null)
        {
            return (null, error);
        }
        var id = InsertJob("bookmarks", null, new JsonObject { ["path"] = sourcePath, ["total"] = urls.Count });
        _bookmarkUrls[id] = urls;
        Launch(id, RunBookmarksImportAsync);
        return (GetRow(id), null);
    }

    private readonly ConcurrentDictionary<string, List<string>> _bookmarkUrls = new();

    private async Task RunBookmarksImportAsync(string id, CancellationToken cancel)
    {
        var urls = _bookmarkUrls.TryRemove(id, out var u) ? u : new List<string>();
        var p = LoadProgress(GetRow(id)!);
        try
        {
            foreach (var url in urls)
            {
                cancel.ThrowIfCancellationRequested();
                try
                {
                    // Already in the library: skip instead of refreshing, so a
                    // resumed import spends API quota only on missing repos.
                    var fullName = RepoUrlParser.Parse(url).FullName;
                    if (_repos.FindByFullName(fullName) is not null)
                    {
                        p.Updated += 1;
                        p.Processed += 1;
                        continue;
                    }
                    var result = await _repos.SaveRepoAsync(url);
                    if (result["outcome"]?.GetValue<string>() == "created") p.Added += 1; else p.Updated += 1;
                }
                catch (Exception ex) when (ex is not OperationCanceledException)
                {
                    p.Failed += 1;
                    p.PushFailure(url, ex is ServiceException se ? se.Code : ex is GitHubException ge ? ge.Code : "error");
                }
                p.Processed += 1;
                if (p.Processed % 5 == 0)
                {
                    Flush(id, p);
                }
            }
            Flush(id, p, u2 =>
            {
                u2.Set("status", "done");
                u2.Set("finished_at", Now());
            });
        }
        catch (OperationCanceledException)
        {
            Flush(id, p, u2 =>
            {
                u2.Set("status", "cancelled");
                u2.Set("finished_at", Now());
            });
        }
        await Task.CompletedTask;
    }

    /// <summary>Resume a failed or cancelled stars import from its last committed page.</summary>
    public JsonObject? ResumeJob(string id)
    {
        var job = GetRow(id);
        if (job is null)
        {
            return null;
        }
        if (job["status"]!.GetValue<string>() == "running" || job["type"]!.GetValue<string>() != "stars")
        {
            return job;
        }
        Update(id, u =>
        {
            u.Set("status", "running");
            u.Set("error", null);
            u.Set("finished_at", null);
        });
        Launch(id, RunStarsImportAsync);
        return GetRow(id);
    }

    public bool CancelJob(string id)
    {
        if (_running.TryGetValue(id, out var cts))
        {
            cts.Cancel();
            return true;
        }
        var job = GetRow(id);
        if (job is not null && job["status"]!.GetValue<string>() == "running")
        {
            Update(id, u =>
            {
                u.Set("status", "cancelled");
                u.Set("finished_at", Now());
            });
            return true;
        }
        return false;
    }

    private void Launch(string id, Func<string, CancellationToken, Task> run)
    {
        var cts = new CancellationTokenSource();
        _running[id] = cts;
        Task.Run(async () =>
        {
            try
            {
                await run(id, cts.Token);
            }
            catch (Exception ex)
            {
                Update(id, u =>
                {
                    u.Set("status", "failed");
                    u.Set("error", ex.Message);
                    u.Set("finished_at", Now());
                });
            }
            finally
            {
                _running.TryRemove(id, out _);
                cts.Dispose();
            }
        });
    }

    private sealed class Progress
    {
        public int NextPage;
        public int Processed;
        public int Added;
        public int Updated;
        public int Failed;
        public readonly List<(string Subject, string Code)> Failures = new();

        public void PushFailure(string subject, string code)
        {
            if (Failures.Count < FailureListCap)
            {
                Failures.Add((subject, code));
            }
        }

        public string FailuresJson() => JsonSerializer.Serialize(Failures.Select(f => new { subject = f.Subject, code = f.Code }));
    }

    private Progress LoadProgress(JsonObject job)
    {
        var failures = new List<(string, string)>();
        if (job["failures"] is JsonArray arr)
        {
            foreach (var f in arr)
            {
                failures.Add((f!["subject"]!.GetValue<string>(), f["code"]!.GetValue<string>()));
            }
        }
        var p = new Progress
        {
            NextPage = (int)job["nextPage"]!.GetValue<long>(),
            Processed = (int)job["processed"]!.GetValue<long>(),
            Added = (int)job["added"]!.GetValue<long>(),
            Updated = (int)job["updated"]!.GetValue<long>(),
            Failed = (int)job["failed"]!.GetValue<long>(),
        };
        p.Failures.AddRange(failures);
        return p;
    }

    private void Flush(string id, Progress p, Action<SqliteUpdate>? extra = null) => Update(id, u =>
    {
        u.Set("next_page", p.NextPage);
        u.Set("processed", p.Processed);
        u.Set("added", p.Added);
        u.Set("updated", p.Updated);
        u.Set("failed", p.Failed);
        u.Set("failures", p.FailuresJson());
        extra?.Invoke(u);
    });

    private async Task RunStarsImportAsync(string id, CancellationToken cancel)
    {
        var job = GetRow(id)!;
        var includeReadme = job["options"]?["includeReadme"]?.GetValue<bool>() ?? true;
        var username = job["username"]!.GetValue<string>()!;
        var p = LoadProgress(job);
        var page = p.NextPage;
        var readmeRateLimited = false;

        try
        {
            while (true)
            {
                cancel.ThrowIfCancellationRequested();
                (List<RepoMeta> Items, int? NextPage) pageData;
                try
                {
                    pageData = await _gh.FetchStarredPageAsync(username, page);
                }
                catch (GitHubException ex) when (ex.Code == "rate_limited"
                    && ex.RetryAfterMs is not null && ex.RetryAfterMs <= _maxRateLimitWaitMs)
                {
                    // Honor the server's wait time, but only when bounded.
                    await Task.Delay((int)ex.RetryAfterMs.Value + 250, cancel);
                    pageData = await _gh.FetchStarredPageAsync(username, page);
                }
                if (pageData.Items.Count == 0)
                {
                    break;
                }

                foreach (var meta in pageData.Items)
                {
                    cancel.ThrowIfCancellationRequested();
                    try
                    {
                        string? readme = null;
                        var readmeTruncated = false;
                        var readmeProvided = false;
                        if (includeReadme && !readmeRateLimited)
                        {
                            readmeProvided = true;
                            try
                            {
                                var readmeResult = await _gh.FetchReadmeAsync(meta.Owner, meta.Name);
                                if (readmeResult is { } r)
                                {
                                    readme = r.Text;
                                    readmeTruncated = r.Truncated;
                                }
                            }
                            catch (GitHubException ex)
                            {
                                p.Failed += 1;
                                p.PushFailure(meta.FullName, $"readme: {ex.Code}");
                                if (ex.Code == "rate_limited")
                                {
                                    readmeRateLimited = true;
                                    readmeProvided = false;
                                    p.PushFailure(meta.FullName, "readme: remaining README fetches skipped (rate limit)");
                                }
                            }
                        }
                        var result = _repos.UpsertSource(meta, readmeProvided ? readme : null, readmeProvided, readmeTruncated, starred: true);
                        lock (_store.Sync)
                        {
                            using var seen = _store.Conn.CreateCommand();
                            seen.CommandText = "INSERT OR IGNORE INTO job_seen (job_id, github_id) VALUES ($job, $gid)";
                            seen.Parameters.AddWithValue("$job", id);
                            seen.Parameters.AddWithValue("$gid", meta.GithubId);
                            seen.ExecuteNonQuery();
                        }
                        if (result.Created) p.Added += 1; else p.Updated += 1;
                    }
                    catch (Exception ex) when (ex is not OperationCanceledException)
                    {
                        p.Failed += 1;
                        p.PushFailure(meta.FullName, ex is GitHubException ge ? ge.Code : "error");
                    }
                    p.Processed += 1;
                }
                p.NextPage = page + 1;
                Flush(id, p);
                if (pageData.NextPage is null)
                {
                    break;
                }
                page = pageData.NextPage.Value;
            }

            // The listing completed: anything still flagged as starred but
            // absent from this run was unstarred upstream. Flag, never delete.
            lock (_store.Sync)
            {
                using var reconcile = _store.Conn.CreateCommand();
                reconcile.CommandText = """
                    UPDATE repos SET starred_upstream = 0
                    WHERE starred_upstream = 1
                      AND github_id NOT IN (SELECT github_id FROM job_seen WHERE job_id = $job)
                    """;
                reconcile.Parameters.AddWithValue("$job", id);
                reconcile.ExecuteNonQuery();
            }
            Flush(id, p, u =>
            {
                u.Set("status", "done");
                u.Set("finished_at", Now());
            });
        }
        catch (OperationCanceledException)
        {
            Flush(id, p, u =>
            {
                u.Set("status", "cancelled");
                u.Set("finished_at", Now());
            });
        }
    }

    private async Task RunRefreshAllAsync(string id, CancellationToken cancel)
    {
        var job = GetRow(id)!;
        var p = LoadProgress(job);
        List<long> ids;
        lock (_store.Sync)
        {
            using var cmd = _store.Conn.CreateCommand();
            cmd.CommandText = "SELECT id FROM repos ORDER BY id";
            using var reader = cmd.ExecuteReader();
            ids = new List<long>();
            while (reader.Read())
            {
                ids.Add(reader.GetInt64(0));
            }
        }
        try
        {
            foreach (var repoId in ids)
            {
                cancel.ThrowIfCancellationRequested();
                try
                {
                    await _repos.RefreshRepoAsync(repoId);
                    p.Updated += 1;
                }
                catch (Exception ex) when (ex is not OperationCanceledException)
                {
                    p.Failed += 1;
                    p.PushFailure($"#{repoId}", ex is ServiceException se ? se.Code : ex is GitHubException ge ? ge.Code : "error");
                }
                p.Processed += 1;
                if (p.Processed % 10 == 0)
                {
                    Flush(id, p);
                }
            }
            Flush(id, p, u =>
            {
                u.Set("status", "done");
                u.Set("finished_at", Now());
            });
        }
        catch (OperationCanceledException)
        {
            Flush(id, p, u =>
            {
                u.Set("status", "cancelled");
                u.Set("finished_at", Now());
            });
        }
        await Task.CompletedTask;
    }
}

File: server/jobs.js
// Background jobs: GitHub Stars import and refresh-all. Jobs run in-process
// (single-user local app), persist progress in `import_jobs` for resume, and
// support cancellation between units of work.

import crypto from 'node:crypto';
import { GitHubError } from './github.js';
import { upsertSource, refreshRepo } from './repos.js';

const now = () => new Date().toISOString();
const FAILURE_LIST_CAP = 100;
const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

export function createJobManager({ db, gh, maxRateLimitWaitMs = 90_000, sleepImpl = sleep } = {}) {
  const running = new Map(); // job id -> { cancel: boolean }

  const getRow = (id) => db.prepare('SELECT * FROM import_jobs WHERE id = ?').get(id);

  function toPublic(row) {
    if (!row) return null;
    return {
      id: row.id,
      type: row.type,
      username: row.username,
      status: row.status,
      options: JSON.parse(row.options || '{}'),
      nextPage: row.next_page,
      processed: row.processed,
      added: row.added,
      updated: row.updated,
      failed: row.failed,
      failures: JSON.parse(row.failures || '[]'),
      error: row.error,
      createdAt: row.created_at,
      finishedAt: row.finished_at,
    };
  }

  function update(id, fields) {
    const keys = Object.keys(fields);
    const sql = `UPDATE import_jobs SET ${keys.map((k) => `${k} = ?`).join(', ')} WHERE id = ?`;
    db.prepare(sql).run(...keys.map((k) => (typeof fields[k] === 'object' ? JSON.stringify(fields[k]) : fields[k])), id);
  }

  function pushFailure(failures, subject, code) {
    if (failures.length < FAILURE_LIST_CAP) failures.push({ subject, code });
  }

  // Honor server rate-limit information: wait only when the reported wait is
  // bounded and reasonable; otherwise stop the job with a clear error instead
  // of hammering the API.
  async function withRateLimitGate(err) {
    if (!(err instanceof GitHubError) || err.code !== 'rate_limited') return false;
    const waitMs = err.retryAfterMs;
    if (waitMs !== null && waitMs <= maxRateLimitWaitMs) {
      await sleepImpl(waitMs + 250);
      return true;
    }
    return false;
  }

  async function runStarsImport(id) {
    const state = running.get(id);
    const job = getRow(id);
    const { includeReadme = true } = JSON.parse(job.options || '{}');
    let { username, next_page: page, processed, added, updated, failed } = job;
    let failures = JSON.parse(job.failures || '[]');
    let readmeRateLimited = false;

    const flush = (extra = {}) => update(id, {
      next_page: page, processed, added, updated, failed, failures, ...extra,
    });

    try {
      while (true) {
        if (state.cancel) {
          flush({ status: 'cancelled', finished_at: now() });
          return;
        }
        let pageData;
        try {
          pageData = await gh.fetchStarredPage(username, page);
        } catch (err) {
          if (await withRateLimitGate(err)) {
            pageData = await gh.fetchStarredPage(username, page);
          } else {
            throw err;
          }
        }
        if (pageData.items.length === 0) break;

        for (const meta of pageData.items) {
          if (state.cancel) {
            flush({ status: 'cancelled', finished_at: now() });
            return;
          }
          try {
            let readme;
            let readmeTruncated = 0;
            if (includeReadme && !readmeRateLimited) {
              try {
                const result = await gh.fetchReadme(meta.owner, meta.name);
                if (result) {
                  readme = result.text;
                  readmeTruncated = result.truncated ? 1 : 0;
                }
              } catch (err) {
                failed += 1;
                pushFailure(failures, meta.fullName, `readme: ${err.code || 'error'}`);
                if (err instanceof GitHubError && err.code === 'rate_limited') {
                  readmeRateLimited = true;
                  pushFailure(failures, meta.fullName, 'readme: remaining README fetches skipped (rate limit)');
                }
              }
            }
            const result = upsertSource(db, meta, { readme, readmeTruncated, starred: 1 });
            db.prepare('INSERT OR IGNORE INTO job_seen (job_id, github_id) VALUES (?, ?)').run(id, meta.githubId);
            if (result.created) added += 1; else updated += 1;
          } catch (err) {
            failed += 1;
            pushFailure(failures, meta.fullName, err.code || 'error');
          }
          processed += 1;
        }
        flush();
        if (!pageData.nextPage) break;
        page = pageData.nextPage;
      }

      // The listing completed: anything still flagged as starred but absent
      // from this run was unstarred upstream. Flag it, never delete it.
      db.prepare(
        `UPDATE repos SET starred_upstream = 0
         WHERE starred_upstream = 1
           AND github_id NOT IN (SELECT github_id FROM job_seen WHERE job_id = ?)`,
      ).run(id);
      flush({ status: 'done', finished_at: now() });
    } catch (err) {
      flush({ status: 'failed', error: err.message, finished_at: now() });
    } finally {
      running.delete(id);
    }
  }

  async function runRefreshAll(id) {
    const state = running.get(id);
    const job = getRow(id);
    let { processed, failed } = job;
    let updated = job.updated;
    let failures = JSON.parse(job.failures || '[]');
    const ids = db.prepare('SELECT id FROM repos ORDER BY id').all().map((r) => r.id);

    try {
      for (const repoId of ids) {
        if (state.cancel) {
          update(id, { status: 'cancelled', processed, updated, failed, failures, finished_at: now() });
          return;
        }
        try {
          await refreshRepo(db, gh, repoId);
          updated += 1;
        } catch (err) {
          failed += 1;
          pushFailure(failures, `#${repoId}`, err.code || 'error');
        }
        processed += 1;
        if (processed % 10 === 0) update(id, { processed, updated, failed, failures });
      }
      update(id, { status: 'done', processed, updated, failed, failures, finished_at: now() });
    } catch (err) {
      update(id, { status: 'failed', processed, updated, failed, failures, error: err.message, finished_at: now() });
    } finally {
      running.delete(id);
    }
  }

  function insertJob(type, { username = null, options = {} } = {}) {
    const id = crypto.randomUUID();
    db.prepare(
      `INSERT INTO import_jobs (id, type, username, status, options, created_at)
       VALUES (?, ?, ?, 'running', ?, ?)`,
    ).run(id, type, username, JSON.stringify(options), now());
    return id;
  }

  return {
    startStarsImport({ username, includeReadme = true } = {}) {
      const id = insertJob('stars', { username, options: { includeReadme } });
      running.set(id, { cancel: false });
      setImmediate(() => runStarsImport(id));
      return toPublic(getRow(id));
    },

    startRefreshAll() {
      const id = insertJob('refresh_all');
      running.set(id, { cancel: false });
      setImmediate(() => runRefreshAll(id));
      return toPublic(getRow(id));
    },

    /** Resume a failed or cancelled stars import from its last committed page. */
    resumeJob(id) {
      const job = getRow(id);
      if (!job) return null;
      if (job.status === 'running') return toPublic(job);
      if (job.type !== 'stars') return toPublic(job);
      update(id, { status: 'running', error: null, finished_at: null });
      running.set(id, { cancel: false });
      setImmediate(() => runStarsImport(id));
      return toPublic(getRow(id));
    },

    cancelJob(id) {
      const state = running.get(id);
      if (state) {
        state.cancel = true;
        return true;
      }
      const job = getRow(id);
      if (job && job.status === 'running') {
        update(id, { status: 'cancelled', finished_at: now() });
        return true;
      }
      return false;
    },

    getJob: (id) => toPublic(getRow(id)),
    listJobs: () => db.prepare('SELECT * FROM import_jobs ORDER BY created_at DESC LIMIT 20').all().map(toPublic),
  };
}
